File: 0046-permutations/0046-permutations.java
class Solution {
    ArrayList<List<Integer>> ans = new ArrayList<>();
    private void backtrack(int[] nums , ArrayList<Integer> current , boolean[] used){
        if(current.size() == nums.length){
            ans.add(new ArrayList<>(current));
            return;
        }

        for(int i = 0 ; i < nums.length ; i++){
            if(used[i]) continue;

            current.add(nums[i]);

            used[i] = true;

            backtrack(nums, current , used);

            current.remove(current.size() - 1);

            used[i] = false;
        }
    }
    public List<List<Integer>> permute(int[] nums) {
        backtrack(nums , new ArrayList<>() , new boolean[nums.length]);
        return ans;
    }
}

File: 0046-permutations/0046-permutations.py
from typing import List

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans = []

        def backtrack(path):
            if len(path) == len(nums):
                ans.append(path[:]) # copy the path
                return

            for num in nums:
                if num in path:
                    continue

                path.append(num)
                backtrack(path)
                path.pop()
        
            return path

        backtrack([])

        return ans